File: webflux-dialogue/src/main/java/com/study/webflux/rag/application/monitoring/aop/MonitoredPipelineAspect.java
package com.study.webflux.rag.application.monitoring.aop;

import lombok.extern.slf4j.Slf4j;

import org.springframework.stereotype.Component;

import com.study.webflux.rag.application.monitoring.context.PipelineContext;
import com.study.webflux.rag.application.monitoring.monitor.DialoguePipelineMonitor;
import com.study.webflux.rag.application.monitoring.monitor.DialoguePipelineTracker;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Slf4j
@Aspect
@Component
public class MonitoredPipelineAspect {

	private final DialoguePipelineMonitor pipelineMonitor;

	public MonitoredPipelineAspect(DialoguePipelineMonitor pipelineMonitor) {
		this.pipelineMonitor = pipelineMonitor;
	}

	@Around("@annotation(monitoredPipeline)")
	public Object wrapPipeline(ProceedingJoinPoint joinPoint,
		MonitoredPipeline monitoredPipeline) throws Throwable {
		Object result = joinPoint.proceed();

		if (!(result instanceof Mono<?> || result instanceof Flux<?>)) {
			return result;
		}

		String inputText = resolveInputText(joinPoint.getArgs(), monitoredPipeline);
		DialoguePipelineTracker tracker = pipelineMonitor.create(inputText);

		if (result instanceof Mono<?> mono) {
			Mono<?> withContext = mono.contextWrite(ctx -> PipelineContext.withTracker(ctx,
				tracker));
			return tracker.attachLifecycle(withContext);
		}

		Flux<?> flux = (Flux<?>) result;
		Flux<?> withContext = flux.contextWrite(ctx -> PipelineContext.withTracker(ctx, tracker));
		return tracker.attachLifecycle(withContext);
	}

	private String resolveInputText(Object[] args, MonitoredPipeline monitoredPipeline) {
		int index = monitoredPipeline.inputArgIndex();
		if (args == null || index < 0 || index >= args.length) {
			log.debug("Monitored pipeline input arg index {} is out of bounds.", index);
			return "";
		}
		Object candidate = args[index];
		if (candidate == null) {
			return "";
		}
		if (candidate instanceof String text) {
			return text;
		}
		return String.valueOf(candidate);
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/application/monitoring/aop/MonitoredPipelineAspect.kt
package com.miyou.app.application.monitoring.aop

import com.miyou.app.application.monitoring.context.PipelineContext
import com.miyou.app.application.monitoring.monitor.DialoguePipelineMonitor
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@Aspect
@Component
class MonitoredPipelineAspect(
    private val pipelineMonitor: DialoguePipelineMonitor,
) {
    private val logger = LoggerFactory.getLogger(MonitoredPipelineAspect::class.java)

    @Around("@annotation(monitoredPipeline)")
    fun wrapPipeline(
        joinPoint: ProceedingJoinPoint,
        monitoredPipeline: MonitoredPipeline,
    ): Any {
        val result = joinPoint.proceed()

        if (result !is Mono<*> && result !is Flux<*>) {
            return result
        }

        val inputText = resolveInputText(joinPoint.args, monitoredPipeline)
        val tracker = pipelineMonitor.create(inputText)

        return if (result is Mono<*>) {
            val withContext =
                result.contextWrite { context ->
                    PipelineContext.withTracker(context, tracker)
                }
            tracker.attachLifecycle(withContext)
        } else {
            val flux = result as Flux<*>
            val withContext =
                flux.contextWrite { context ->
                    PipelineContext.withTracker(context, tracker)
                }
            tracker.attachLifecycle(withContext)
        }
    }

    private fun resolveInputText(
        args: Array<Any>,
        monitoredPipeline: MonitoredPipeline,
    ): String {
        val index = monitoredPipeline.inputArgIndex
        if (args.isEmpty() || index < 0 || index >= args.size) {
            logger.debug("Monitored pipeline input arg index {} is out of bounds.", index)
            return ""
        }

        val candidate = args[index]
        return when {
            candidate == null -> ""
            candidate is String -> candidate
            else -> candidate.toString()
        }
    }
}
